### PR TITLE
Migrating from U2F to WebAuthn for 2FA

### PR DESCRIPTION
**Pull request by FreddleSpl0it**
### Description
The Google Chrome browser only supports the legacy U2F standard until February 2022 and therefore the current 2FA implementation will no longer work. Migrating to the new WebAuthn standard will solve this issue.

### This PR closes the following Issues
Closes [#4421](https://github.com/mailcow/mailcow-dockerized/issues/4421)
fixes some issues on android with the current fido2 login

### CODE REVIEW

**Comment by DerLinkman** [comment]
Looks good so far. Before we merge I would like to get a review from @mkuron as well as another Live Test (a mailcow paid support customer was asking about it).

**Comment by mkuron** [comment]
I have looked at it a few days ago and it looks reasonable. However, I am not an expert on Webauthn at all, so we should definitely get another review from someone who is actually familiar with it.

**Comment by DerLinkman** [comment]
As i mentioned earlier tomorrow (hopefully) i will test this patch on a running machine.
Android and Apple WebAuthn works as expected (or even better) with the new Version.

**Comment by DerLinkman** [comment]
Test failed! It did´nt worked. This key was used:

https://www.token2.swiss/shop/product/token2-t2f2-alu-fido2-u2f-and-totp-security-key

**Comment by FreddleSpl0it** [comment]
I have disabled rootCA validation by default because of this comment [lbuchs/WebAuthn/issues/31.](https://github.com/lbuchs/WebAuthn/issues/31#issuecomment-763597036) This function can still be enabled via mailcow.conf.

File: generate_config.sh [hunk 1]
```diff
@@ -344,22 +344,10 @@ DOVECOT_MASTER_PASS=
...
-#   googleHardware certified
-#   microsoftTpmCollection certified
-#   huawei certified
-#   trustkey certified
-#   bsi certified
-WEBAUTHN_DISABLE_ROOTCA=n
+# Enable webauthn device manufacturer verification
+# After setting WEBAUTHN_RESPECT_ROOTCA=y only devices from trusted manufacturers are allowed
+# root certificates can be placed for validation under mailcow-dockerized/data/web/inc/lib/WebAuthn/rootCertificates
+WEBAUTHN_RESPECT_ROOTCA=n
```
Comment (andryyy) [inline generate_config.sh:350]: Automatically enabled?

File: generate_config.sh [hunk 2]
```diff
@@ -344,22 +344,10 @@ DOVECOT_MASTER_PASS=
...
-#   globalSign certified
-#   googleHardware certified
-#   microsoftTpmCollection certified
-#   huawei certified
-#   trustkey certified
-#   bsi certified
-WEBAUTHN_DISABLE_ROOTCA=n
+# Enable webauthn device manufacturer verification
+# After setting WEBAUTHN_RESPECT_ROOTCA=y only devices from trusted manufacturers are allowed
+# root certificates can be placed for validation under mailcow-dockerized/data/web/inc/lib/WebAuthn/rootCertificates
```
Comment (andryyy) [inline generate_config.sh:349]: Automatically enabled?!

File: update.sh [hunk 3]
```diff
@@ -307,7 +307,7 @@ CONFIG_ARRAY=(
   "ADDITIONAL_SERVER_NAMES"
   "ACME_CONTACT"
   "WATCHDOG_VERBOSE"
-  "WEBAUTHN_DISABLE_ROOTCA"
+  "WEBAUTHN_RESPECT_ROOTCA"
```
Comment (andryyy) [inline update.sh]: Are there more inconsistencies of respect and disable?

**Comment by andryyy** [comment]
Questions might be already answered. Pretty low on time. Only on mobile, too. Sorry.

**Comment by andryyy** [comment]
We allow for key only logins. Without passwords. If we do this with untrusted root CAs that’s pretty crazy.

**Comment by andryyy** [comment]
Add a link to the docs about how to allow a new CA, but please don’t trust it. I think that’s more than crazy when using 🔑-only logins. :/

**Comment by mkuron** [comment]
Actually, I think the correct thing to do may be to ignore the root CA. It is only needed to determine if a device is from a specific vendor. I don‘t think it‘s part of the public-key authentication. See https://developers.yubico.com/U2F/Attestation_and_Metadata/ and https://developers.yubico.com/WebAuthn/Concepts/Securing_WebAuthn_with_Attestation.html for what Yubico says about this.

**Comment by FreddleSpl0it** [comment]
I can undo the changes. It was an idea that came to me while reading this thread https://github.com/lbuchs/WebAuthn/issues/31#issuecomment-780401348, since the author of the WebAuthn library himself says it is not necessary

**Comment by DerLinkman** [comment]
As i mentioned earlier i tested this patch on a customer machine which had a Third Party FIDO Key (https://www.token2.swiss/shop/product/token2-t2f2-alu-fido2-u2f-and-totp-security-key) registered before as U2F. But as soon he wanted to register the same Key with WebAuthn he wasn´t able to because "Error invalid root Certificate - (reload browser if the error persists)" (and yes on a different browser, the same problem), which is the exact problem i guess.

**Comment by andryyy** [comment]
My concern was yet to limit WebAuthn to trusted devices. Not really about the usefulness of the patch itself.

**Comment by andryyy** [comment]
Thanks for all the input, guys. 🙏

**Comment by DerLinkman** [comment]
That's understandable 😊
I only want to say that the original planned version (so respecting the Root CAs) probably won't work for users which don't have a Key from a manufacturer included into the CAs.
________________________________

**Comment by DerLinkman** [comment]
Ok the new tests claims that it is working successfully (same customer and same key as yesterday)

The new Variable in the mailcow.conf is checking if the Webauthn should ignore the certificates or not.

I already started working on the documentation: 
https://github.com/DerLinkman/mailcow-dockerized-docs/blob/WebAuthn-TFA-Documentation/docs/u_e-mailcow_ui-tfa.md

So @andryyy and or @mkuron what do you think about it?

**Comment by andryyy** [comment]
Hey @FreddleSpl0it, sorry for being a jerk, just figured something out internally.

Thanks for your contribution and help. I got something completely wrong.

**Pull request merged**